### PR TITLE
Add Circle `contains()`

### PR DESCRIPTION
**Pull request by itzpr3d4t0r**
This PR adds the Circle.contains() method for checking when a shape (Rect/Frect/Circle for now) or point falls completely inside the circle.

## Credits

**Geometry Project:**
For code, docs and tests:
@novialriptide @Emc2356 @itzpr3d4t0r @ScriptLineStudios @avaxar @Matiiss @newpaxonian @maqa41 @blankRiot96
Also thanks to @Starbuck5 for kickstarting the idea and the occasional help!

**Functionality added in this PR**
- `circle.c`
https://github.com/pygame-community/pygame-geometry/blame/main/src_c/circle.c
Credits to @Emc2356 @itzpr3d4t0r @novialriptide
- `geometry.pyi`
https://github.com/pygame-community/pygame-geometry/blame/main/geometry.pyi
Credits to @Emc2356 @itzpr3d4t0r @novialriptide @ScriptLineStudios @avaxar 
- `geometry_test.py`
https://github.com/pygame-community/pygame-geometry/blame/main/test/test_circle.py
Credits to @itzpr3d4t0r @novialriptide
- `geometry.rst`
https://github.com/pygame-community/pygame-geometry/blame/main/docs/circle.rst
Credits to @itzpr3d4t0r

### CODE REVIEW

File: src_c/circle.c [hunk 1]
```diff
@@ -371,6 +371,65 @@ pg_circle_as_frect(pgCircleObject *self, PyObject *_null)
...
+            /*a circle is always contained within itself*/
+            Py_RETURN_TRUE;
+        }
+        double dx, dy, dr;
+
+        dx = temp->x - scirc->x;
+        dy = temp->y - scirc->y;
+        dr = temp->r - scirc->r;
+
+        result = (dx * dx + dy * dy) <= (dr * dr);
```
Comment (ankith26) [inline src_c/circle.c:393]: This algorithm does not seem to take the sign of `dr` into account, which seems wrong.

Like, I am seeing that swapping `temp` and `scirc` in this algorithm would not change the result, which is logically wrong, as `a.contains(b) does not imply `b.contains(a)`

Comment (itzpr3d4t0r) [reply src_c/circle.c]: Thanks a lot for noticing!! It should be fixed now. Tomorrow I'll add another test to account for this.

File: docs/reST/ref/geometry.rst [hunk 2]
```diff
@@ -283,6 +283,29 @@
...
+         | :sl:`test if a shape or point is inside the circle`
+         | :sg:`contains(circle, /) -> bool`
+         | :sg:`contains(rect, /) -> bool`
+         | :sg:`contains((x, y), /) -> bool`
+         | :sg:`contains(vector2, /) -> bool`
+
+         The `contains` method tests whether a given shape or point is completely contained
+         within a `Circle` object. The function takes a single argument which can be a `Circle`,
+         `Rect`, `FRect`, or a point.
+         It returns `True` if the shape or point is completely contained, and `False` otherwise.
```
Comment (itzpr3d4t0r) [inline docs/reST/ref/geometry.rst:297]: Note to self, fix docs, remove use of both "method" and "function", remove direct name citation

File: src_c/circle.c [hunk 3]
```diff
@@ -371,6 +371,66 @@ pg_circle_as_frect(pgCircleObject *self, PyObject *_null)
...
+        SDL_FRect *temp = &pgFRect_AsRect(arg);
+
+        if (pgCollision_CirclePoint(scirc, (double)temp->x, (double)temp->y) &&
+            pgCollision_CirclePoint(scirc, (double)(temp->x + temp->w),
+                                    (double)temp->y) &&
+            pgCollision_CirclePoint(scirc, (double)temp->x,
+                                    (double)(temp->y + temp->h)) &&
+            pgCollision_CirclePoint(scirc, (double)(temp->x + temp->w),
+                                    (double)(temp->y + temp->h))) {
+            result = 1;
```
Comment (MyreMylar) [inline src_c/circle.c:419]: Feels like this duplication could be condensed - maybe a macro is best:

```
#define RECT_CIRCLE_CONTAINS(rect, circle, result)                              \
    if (pgCollision_CirclePoint(scirc, (double)rect->x, (double)rect->y) &&     \
            pgCollision_CirclePoint(circle, (double)(rect->x + rect->w),        \
                                    (double)rect->y) &&                         \
            pgCollision_CirclePoint(circle, (double)rect->x,                    \
                                    (double)(rect->y + rect->h)) &&             \
            pgCollision_CirclePoint(circle, (double)(rect->x + rect->w),        \
                                    (double)(rect->y + rect->h))) {             \
            result = 1;                                                         \
        }   
...
    else if (pgRect_Check(arg)) {
        SDL_Rect *temp = &pgRect_AsRect(arg);

        RECT_CIRCLE_CONTAINS(temp, scirc, result)
    }
    else if (pgFRect_Check(arg)) {
        SDL_FRect *temp = &pgFRect_AsRect(arg);

        RECT_CIRCLE_CONTAINS(temp, scirc, result)
```

**Review by ankith26 — APPROVED** [review]
LGTM, thanks for the PR :partying_face:

File: src_c/circle.c [hunk 4]
```diff
@@ -371,6 +371,61 @@ pg_circle_as_frect(pgCircleObject *self, PyObject *_null)
...
+
+static PyObject *
+pg_circle_contains(pgCircleObject *self, PyObject *arg)
+{
+    int result = 0;
+    pgCircleBase *scirc = &self->circle;
+    double x, y;
+
+    if (pgCircle_Check(arg)) {
+        pgCircleBase *temp = &pgCircle_AsCircle(arg);
```
Comment (oddbookworm) [inline src_c/circle.c:393]: For each of these, is `temp` the best name? I know that's nitpicky, but it just doesn't feel right to me because that's basically a recasting of the passed in arg

File: test/geometry_test.py [hunk 5]
```diff
@@ -1149,6 +1149,112 @@ def assert_approx_equal(circle1, circle2, eps=1e-12):
...
+    def test_contains_circle(self):
+        """Ensures that the contains method correctly determines if a circle is
+        contained within the circle"""
+        c = Circle(10, 10, 4)
+        c2 = Circle(10, 10, 2)
+        c3 = Circle(100, 100, 5)
+        c4 = Circle(16, 10, 7)
+
+        # self
+        self.assertTrue(c.contains(c))
```
Comment (oddbookworm) [inline test/geometry_test.py:1191]: I'd like to see an edge case where you create an identical circle to `c` and then check if `c` contains that identical circle. It should, but it should be tested because it's an edge case

File: test/geometry_test.py [hunk 6]
```diff
@@ -1149,6 +1149,112 @@ def assert_approx_equal(circle1, circle2, eps=1e-12):
...
+        self.assertFalse(c.contains(c4))
+
+        # bigger circle not contained in smaller circle
+        self.assertFalse(c2.contains(c))
+        self.assertFalse(c3.contains(c4))
+
+    def test_contains_point(self):
+        """Ensures that the contains method correctly determines if a point is
+        contained within the circle"""
+        c = Circle(10, 10, 4)
```
Comment (oddbookworm) [inline test/geometry_test.py:1209]: I'd like to see a case where the point is on the circumference of the circle, just to have that edge case tested (quite literally edge case lol)

File: test/geometry_test.py [hunk 7]
```diff
@@ -1149,6 +1149,112 @@ def assert_approx_equal(circle1, circle2, eps=1e-12):
...
+        self.assertTrue(c.contains(p1v))
+
+        # not contained point
+        self.assertFalse(c.contains(p2v))
+        self.assertFalse(c.contains(p3v))
+
+    def test_contains_rect_frect(self):
+        """Ensures that the contains method correctly determines if a rect is
+        contained within the circle"""
+        c = Circle(0, 0, 10)
```
Comment (oddbookworm) [inline test/geometry_test.py:1235]: Similar to the other tests, I'd like to see one edge case tested. Specifically, where all 4 corners of the rect are on the perimeter of the circle

**Review by oddbookworm — DISMISSED** [review]
This almost looks ready to merge IMO. I just have a few small comments, 3 of which are just for making the tests slightly more robust

**Review by MyreMylar — APPROVED** [review]
OK, LGTM 👍
